### PR TITLE
Support SSRC conditions

### PR DESCRIPTION
**Pull request by erikeldridge**
Add support for condition evaluation.

Note this is based on https://github.com/firebase/firebase-admin-node/pull/2496.

### Discussion

Working with @lahirumaramba and @trekforever.

### Testing

Ran npm test and all tests pass.

Functionally tested using a local server.

### CODE REVIEW

File: src/remote-config/remote-config-api.ts [hunk 1]
```diff
@@ -56,10 +56,10 @@ export interface RemoteConfigCondition {
 
 /**
  * Represents a Remote Config condition in the dataplane.
- * A condition targets a specific group of users. A list of these conditions make up
- * part of a Remote Config template.
+ * A condition targets a specific group of users. A list of these conditions
+ * make up part of a Remote Config template.
  */
-export interface RemoteConfigServerCondition {
+export interface RemoteConfigServerNamedCondition {
```
Comment (erikeldridge) [inline src/remote-config/remote-config-api.ts:62]: @lahirumaramba I see some interfaces with the RemoteConfig prefix and others without. Is there a guideline for what needs the prefix?

**Comment by erikeldridge** [comment]
@lahirumaramba Any idea what could cause the "Continuous Integration / build (14.x)" failure? I installed Node 14 on my machine and ran `npm install` and `npm run test:unit` and everything worked.

File: src/remote-config/remote-config-condition-evaluator-internal.ts [hunk 2]
```diff
@@ -0,0 +1,173 @@
+/*!
+ * Copyright 2020 Google Inc.
```
Comment (jenh) [inline src/remote-config/remote-config-condition-evaluator-internal.ts:2]: Copyright 2024?

File: test/unit/remote-config/remote-config-condition-evaluator.spec.ts [hunk 3]
```diff
@@ -0,0 +1,364 @@
+/*!
+ * Copyright 2020 Google Inc.
```
Comment (jenh) [inline test/unit/remote-config/remote-config-condition-evaluator.spec.ts:2]: 2024?

File: src/remote-config/remote-config-api.ts [hunk 4]
```diff
@@ -56,10 +56,10 @@ export interface RemoteConfigCondition {
 
 /**
  * Represents a Remote Config condition in the dataplane.
- * A condition targets a specific group of users. A list of these conditions make up
- * part of a Remote Config template.
+ * A condition targets a specific group of users. A list of these conditions
+ * make up part of a Remote Config template.
```
Comment (jenh) [inline src/remote-config/remote-config-api.ts:60]:
```suggestion
 * comprise part of a Remote Config template.
```

File: test/unit/remote-config/remote-config-condition-evaluator.spec.ts [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+          }
+        };
+        const evaluator = new RemoteConfigConditionEvaluator();
+        const truthyAssignments = evaluateRandomAssignments(percentCondition, 100_000, evaluator);
+        const range = 500; // standard deviation + some tolerance
+        expect(truthyAssignments).to.be.greaterThanOrEqual(50000 - range);
+        expect(truthyAssignments).to.be.lessThanOrEqual(50000 + range);
+      });
+
+      // Returns the number of assignments which evaluate to true for the specified percent condition
```
Comment (jenh) [inline test/unit/remote-config/remote-config-condition-evaluator.spec.ts:341]:
```suggestion
      // Returns the number of assignments which evaluate to true for the specified percent condition.
```

File: test/unit/remote-config/remote-config-condition-evaluator.spec.ts [hunk 6]
```diff
@@ -0,0 +1,364 @@
...
+      // This method randomly generates the ids for each assignment for this purpose
```
Comment (jenh) [inline test/unit/remote-config/remote-config-condition-evaluator.spec.ts:342]:
```suggestion
      // This method randomly generates the ids for each assignment for this purpose.
```

File: src/remote-config/remote-config-condition-evaluator-internal.ts (hunk 2 above)
Comment (erikeldridge) [reply src/remote-config/remote-config-condition-evaluator-internal.ts]: Done

File: test/unit/remote-config/remote-config-condition-evaluator.spec.ts (hunk 3 above)
Comment (erikeldridge) [reply test/unit/remote-config/remote-config-condition-evaluator.spec.ts]: Done

**Review by jenh — APPROVED** [review]

**Comment by erikeldridge** [comment]
> @lahirumaramba Any idea what could cause the "Continuous Integration / build (14.x)" failure? I installed Node 14 on my machine and ran `npm install` and `npm run test:unit` and everything worked.

Scratch that. I can repro by:

1. use Node 14 (`nvm use 14`)
2. Checkout feature branch
3. run the `npm ci` commend mentioned in the error
4. Observe error

For comparison:

1. As above
2. Checkout master branch
3. As above
4. Observe success

I'm _guessing_ farmhash might be pulling in a dep that's unsupported on Node 14, since that's the only new dep.

**Comment by erikeldridge** [comment]
> > @lahirumaramba Any idea what could cause the "Continuous Integration / build (14.x)" failure? I installed Node 14 on my machine and ran `npm install` and `npm run test:unit` and everything worked.

> I'm _guessing_ farmhash might be pulling in a dep that's unsupported on Node 14, since that's the only new dep.

Double scratch that. I see `@fastify/busboy` is an existing dep of the Admin SDK. Is this a known issue?

File: src/remote-config/remote-config-api.ts (hunk 1 above)
Comment (lahirumaramba) [reply src/remote-config/remote-config-api.ts:62]: I can't recall a reason. Looking at this proposal ([go/admin-sdk-remote-config](http://go/admin-sdk-remote-config)) from 2020 it looks like the types that were added later are missing the prefix.... maybe we weren't strict about the prefix in follow up proposals?

**Comment by erikeldridge** [comment]
> > > @lahirumaramba Any idea what could cause the "Continuous Integration / build (14.x)" failure? I installed Node 14 on my machine and ran `npm install` and `npm run test:unit` and everything worked.
> 
> > I'm _guessing_ farmhash might be pulling in a dep that's unsupported on Node 14, since that's the only new dep.
> 
> Double scratch that. I see [`@fastify/busboy` is an existing dep of the Admin SDK](https://github.com/firebase/firebase-admin-node/blob/05203aeef7c0fb6e24a85a332c064c53813e3a81/package.json#L199). Is this a known issue?

@lahirumaramba discussed offline. For future ref, he suggested deleting and regenerating package-lock.json, which fixed the issue.

File: src/remote-config/remote-config-api.ts (hunk 1 above)
Comment (erikeldridge) [reply src/remote-config/remote-config-api.ts]: I was just reviewing a change for the REST API and I think what happened is the TS API naming biased toward the REST API. The REST API has nested protos, but TS interfaces can't be nested. The parent protos had more qualified names than the child protos. When we created interfaces to match the protos, we retained the existing names.

I think the guideline we discussed against product prefixes still holds and will help the TS API be more consistent.

**Review by lahirumaramba — APPROVED** [review]
Thanks! LGTM!

**Pull request merged**